submissions/api.py

```diff
@@ -373,6 +375,44 @@ def get_submissions(student_item_dict, limit=None):
     return SubmissionSerializer(submission_models, many=True).data
 
 
+def get_all_submissions(course_id, item_id, item_type, read_replica=True):
+    """For the given item, get the most recent submission for every student who has submitted.
+
+    This may return a very large result set! It is implemented as a generator for efficiency.
+
+    Args:
+        course_id, item_id, item_type (string): The values of the respective student_item fields
+            to filter the submissions by.
+        read_replica (bool): If true, attempt to use the read replica database.
+            If no read replica is available, use the default database.
+
+    Yields:
+        Dicts representing the submissions with the following fields:
+            student_item
+            attempt_number
+            submitted_at
+            created_at
+            answer
+
+    Raises:
+        Cannot fail unless there's a database error, but may return an empty iterable.
+    """
+    submission_qs = Submission.objects
+    if read_replica:
+        submission_qs = _use_read_replica(submission_qs)
+    query = submission_qs.select_related('student_item').filter(
+        student_item__course_id=course_id,
+        student_item__item_id=item_id,
+        student_item__item_type=item_type,
+    ).order_by('student_item__student_id', '-created_at').iterator()
+
+    # We cannot use SELECT DISTINCT ON because it's PostgreSQL only, so unfortunately
+    # our results will contain every entry of each student, not just the most recent.
+    for unused_student_id, row_iter in itertools.groupby(query, operator.attrgetter('student_item.student_id')):
+        submission = next(row_iter)
+        yield SubmissionSerializer(submission).data
```

Comment:
This change makes the data export block not work for me. Because `SubmissionSerializer(submission).data` looks like this:

```python
{
    'uuid': u'ec26ccac-f8fd-11e4-a87f-080027880ca6',
    'student_item': 2L,
    'attempt_number': 5L,
    'submitted_at': datetime.datetime(2015, 5, 12, 23, 23, 39, tzinfo=<UTC>),
    'created_at': datetime.datetime(2015, 5, 12, 23, 23, 39, tzinfo=<UTC>),
    'answer': u'neither'
}
```

But the latest code in the data export block needs this result to contain `student_id` (it must be in the result, since querying the student_item to look it up each time would be too expensive):

```python
for submission in sub_api.get_all_submissions(course_key_str, block_id, block_type):
    student_id = submission['student_id']
```